contracts/asset-proxy/contracts/src/bridges/Eth2DaiBridge.sol

```diff
@@ -0,0 +1,162 @@
+
+    /// @dev Whether we've granted an allowance to a spender for a token.
+    mapping (address => mapping (address => bool)) private _hasAllowance;
+
+    /// @dev Callback for `IERC20Bridge`. Tries to buy `amount` of
+    ///      `toTokenAddress` tokens by selling the entirety of the opposing asset
+    ///      (DAI or WETH) to the Eth2Dai contract, then transfers the bought
+    ///      tokens to `to`.
+    /// @param toTokenAddress The token to give to `to` (either DAI or WETH).
+    /// @param to The recipient of the bought tokens.
+    /// @param amount Minimum amount of `toTokenAddress` tokens to buy.
+    /// @param bridgeData The abi-encoeded "from" token address.
+    /// @return success The magic bytes if successful.
+    function withdrawTo(
+        address toTokenAddress,
+        address /* from */,
+        address to,
+        uint256 amount,
+        bytes calldata bridgeData
+    )
+        external
+        returns (bytes4 success)
+    {
+        // Decode the bridge data to get the `fromTokenAddress`.
+        (address fromTokenAddress) = abi.decode(bridgeData, (address));
+
+        IEth2Dai exchange = _getEth2DaiContract();
+        // Grant an allowance to the exchange to spend `fromTokenAddress` token.
+        _grantAllowanceForToken(address(exchange), fromTokenAddress);
+
+        // Try to sell all of this contract's `fromTokenAddress` token balance.
+        uint256 boughtAmount = _getEth2DaiContract().sellAllAmount(
+            address(fromTokenAddress),
+            IERC20Token(fromTokenAddress).balanceOf(address(this)),
+            toTokenAddress,
+            amount
+        );
+        // Transfer the converted `toToken`s to `to`.
+        _transferERC20Token(toTokenAddress, to, boughtAmount);
+        return BRIDGE_SUCCESS;
+    }
+
+    /// @dev `SignatureType.Wallet` callback, so that this bridge can be the maker
+    ///      and sign for itself in orders. Always succeeds.
+    /// @return magicValue Magic success bytes, always.
+    function isValidSignature(
+        bytes32,
+        bytes calldata
+    )
+        external
+        view
+        returns (bytes4 magicValue)
+    {
+        return LEGACY_WALLET_MAGIC_VALUE;
+    }
+
+    /// @dev Overridable way to get the eth2dai contract.
+    /// @return exchange The Eth2Dai exchange contract.
+    function _getEth2DaiContract()
+        internal
+        view
+        returns (IEth2Dai exchange)
+    {
+        return IEth2Dai(ETH2DAI_ADDRESS);
+    }
+
+    /// @dev Grants an unlimited allowance to `spender` for `tokenAddress` token,
+    ///      if we haven't done so already.
+    /// @param spender The spender address.
+    /// @param tokenAddress The token address.
+    function _grantAllowanceForToken(
+        address spender,
+        address tokenAddress
+    )
+        private
+    {
+        mapping (address => bool) storage spenderHasAllowance = _hasAllowance[spender];
```

Comment:
What happens when the allowance decreases to less than the required amount, but this is still set to true? 

I think that this check actually just ends up being an extra expense most of the time. I'd prefer just calling `IERC20Token(tokenAddress).approve(spender, uint256(-1))` every single time. If an allowance is already set, then this won't cost much extra gas at all. This also allows us to remove the `_hasAllowance` mapping altogether.

Reply:
I'm OK with this.